plugins/askrene/refine.c

```diff
@@ -0,0 +1,379 @@
+	size_t idx;
+
+	get_scidd(rq->gossmap, flow, i, &scidd);
+
+	/* This should not happen, but simply don't reserve if it does */
+	if (!reserves_add(askrene->reserved, &scidd, &amt, 1)) {
+		plugin_log(rq->plugin, LOG_BROKEN,
+			   "Failed to reserve %s in %s",
+			   fmt_amount_msat(tmpctx, amt),
+			   fmt_short_channel_id_dir(tmpctx, &scidd));
+		return;
+	}
+
+	/* Set capacities entry to 0 so it get_constraints() looks in reserve. */
+	idx = gossmap_chan_idx(rq->gossmap, flow->path[i]);
+	if (idx < tal_count(rq->capacities))
+		rq->capacities[idx] = 0;
+
+	/* Record so destructor will unreserve */
+	tal_arr_expand(&r->scidds, scidd);
+	tal_arr_expand(&r->amounts, amt);
+}
+
+/* We have a basic set of flows, but we need to add fees, and take into
+ * account that "spendable" estimates are for a single HTLC.  This can
+ * push us again over capacity or htlc_maximum_msat.
+ *
+ * We may have to reduce the flow amount in response to these.
+ *
+ * We also check for going below htlc_maximum_msat at this point: this
+ * is unusual and means it's small, so we just remove that flow if
+ * this happens, as we can make it up by buffing up the other flows
+ * (or, it's simply impossible).
+ */
+static const char *constrain_flow(const tal_t *ctx,
+				  struct route_query *rq,
+				  struct flow *flow,
+				  struct reservations *reservations)
+{
+	struct amount_msat msat;
+	int decreased = -1;
+	const char *why_decreased;
+
+	/* Walk backwards, adding fees and testing for htlc_max and
+	 * capacity limits. */
+	msat = flow->delivers;
+	for (int i = tal_count(flow->path) - 1; i >= 0; i--) {
+		const struct half_chan *h = flow_edge(flow, i);
+		struct amount_msat min, max, amount_to_reserve;
+		struct short_channel_id_dir scidd;
+		const char *max_cause;
+
+		/* We can pass constraints due to addition of fees! */
+		get_constraints(rq, flow->path[i], flow->dirs[i], &min, &max);
+		if (amount_msat_less(amount_msat(fp16_to_u64(h->htlc_max)), max)) {
+			max_cause = "htlc_maximum_msat of ";
+			max = amount_msat(fp16_to_u64(h->htlc_max));
+		} else {
+			max_cause = "channel capacity of ";
+		}
+
+		/* If amount is > max, we decrease and add note it in
+		 * case something goes wrong later. */
+		if (amount_msat_greater(msat, max)) {
+			plugin_log(rq->plugin, LOG_DBG,
+				   "Decreased %s to %s%s across %s",
+				   fmt_amount_msat(tmpctx, msat),
+				   max_cause,
+				   fmt_amount_msat(tmpctx, max),
+				   fmt_flows_step_scid(tmpctx, rq, flow, i));
+			msat = max;
+			decreased = i;
+			why_decreased = max_cause;
+		}
+
+		/* Reserve more for local channels if it reduces capacity */
+		get_scidd(rq->gossmap, flow, i, &scidd);
+		if (!amount_msat_add(&amount_to_reserve, msat,
+				     get_additional_per_htlc_cost(rq, &scidd)))
+			abort();
+
+		/* Reserve it, so if the next flow asks about the same channel,
+		   it will see the reduced capacity from this one.  */
+		add_reservation(reservations, rq, flow, i, amount_to_reserve);
```

Comment:
I think this is prematurely adding reservations. This loop is used to figure out how much can we send through, once we know that then it makes sense to apply a reservation. If the intended flow can fully go through then there is no problem here, but if the flow amount is reduced then we end up reserving more than we need.

Reply:
The additional cost depends only on the flows' existence, not its amount.

The issue is if two flows use the same channel.  The first one can get up to spendable, *then* we reserve an additional bit before we deal with the second one.  This is because `spendable` already takes into account the reservation for the first channel.

IOW, we end up over-reserving, but only on the final iteration where we don't use the reservation again?